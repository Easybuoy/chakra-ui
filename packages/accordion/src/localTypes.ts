import { chakra, PropsOf } from "@chakra-ui/system"

export type DivProps = PropsOf<typeof chakra.div>
