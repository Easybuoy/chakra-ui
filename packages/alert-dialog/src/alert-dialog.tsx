import {
  Modal,
  ModalContent,
  ModalContentProps,
  ModalProps,
  ModalBody,
  ModalHeader,
  ModalFooter,
  ModalCloseButton,
  ModalOverlay,
} from "@chakra-ui/modal"
import React, { Ref, forwardRef } from "react"

export interface AlertDialogProps extends Omit<ModalProps, "initialFocusRef"> {
  leastDestructiveRef: ModalProps["initialFocusRef"]
}

export function AlertDialog(props: AlertDialogProps) {
  const { leastDestructiveRef, ...rest } = props
  return <Modal {...rest} initialFocusRef={leastDestructiveRef} />
}

export const AlertDialogContent = forwardRef(function AlertDialogContent(
  props: ModalContentProps,
  ref: Ref<HTMLElement>,
) {
  return <ModalContent ref={ref} role="alertdialog" {...props} />
})

export const AlertDialogBody = ModalBody
export const AlertDialogHeader = ModalHeader
export const AlertDialogFooter = ModalFooter
export const AlertDialogCloseButton = ModalCloseButton
export const AlertDialogOverlay = ModalOverlay
