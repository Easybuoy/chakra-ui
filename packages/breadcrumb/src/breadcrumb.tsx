import {
  chakra,
  forwardRef,
  PropsOf,
  useMultiStyleConfig,
  StylesProvider,
  useStyles,
  SystemProps,
  ThemingProps,
  omitThemingProps,
} from "@chakra-ui/system"
import { cx, getValidChildren, __DEV__ } from "@chakra-ui/utils"
import React, { Ref, cloneElement, ReactNode, ReactElement } from "react"

export type BreadcrumbSeparatorProps = PropsOf<typeof chakra.div> & {
  spacing?: SystemProps["mx"]
}

/**
 * React component that separates each breadcrumb link
 */
export const BreadcrumbSeparator = forwardRef(function BreadcrumbSeparator(
  props: BreadcrumbSeparatorProps,
  ref: Ref<HTMLSpanElement>,
) {
  const { spacing, ...rest } = props

  const styles = useStyles()
  const separatorStyles = {
    mx: spacing,
    ...styles.separator,
  }

  return (
    <chakra.span
      ref={ref}
      role="presentation"
      {...rest}
      __css={separatorStyles}
    />
  )
})

if (__DEV__) {
  BreadcrumbSeparator.displayName = "BreadcrumbSeparator"
}

interface LinkOptions {
  isCurrentPage?: boolean
}

export type BreadcrumbLinkProps = Omit<PropsOf<typeof chakra.a>, "children"> &
  LinkOptions & { children: ReactNode }

/**
 * Breadcrumb link.
 *
 * It renders a `span` when it's the current link. Otherwise,
 * it renders an anchor tag.
 */
export const BreadcrumbLink = forwardRef(function BreadcrumbLink(
  props: BreadcrumbLinkProps,
  ref: Ref<any>,
) {
  const { isCurrentPage, as, className, ...rest } = props
  const styles = useStyles()

  const sharedProps = {
    ref,
    as,
    className: cx("chakra-breadcrumb__link", className),
    ...rest,
  }

  if (isCurrentPage) {
    return <chakra.span aria-current="page" {...sharedProps} />
  }

  return <chakra.a __css={styles.link} {...sharedProps} />
})

if (__DEV__) {
  BreadcrumbLink.displayName = "BreadcrumbLink"
}

type BreadcrumbItemOptions = BreadcrumbProps & {
  isCurrentPage?: boolean
  isLastChild?: boolean
}

export type BreadcrumbItemProps = BreadcrumbItemOptions &
  PropsOf<typeof chakra.li>

/**
 * BreadcrumbItem is used to group a breadcrumb link.
 * It renders a `li` element to denote it belongs to an order list of links.
 * @see Docs https://chakra-ui.com/components/breadcrumbs
 */
export const BreadcrumbItem = forwardRef(function BreadcrumbItem(
  props: BreadcrumbItemProps,
  ref: Ref<HTMLLIElement>,
) {
  const {
    isCurrentPage,
    separator,
    isLastChild,
    spacing,
    children,
    className,
    ...rest
  } = props

  const validChildren = getValidChildren(children)

  const clones = validChildren.map((child) => {
    if (child.type === BreadcrumbLink) {
      return cloneElement(child, {
        isCurrentPage,
      })
    }

    if (child.type === BreadcrumbSeparator) {
      return cloneElement(child, {
        spacing,
        children: child.props.children || separator,
      })
    }

    return child
  })

  const _className = cx("chakra-breadcrumb__list-item", className)

  return (
    <chakra.li
      ref={ref}
      className={_className}
      {...rest}
      __css={{
        display: "inline-flex",
        alignItems: "center",
      }}
    >
      {clones}
      {!isLastChild && (
        <BreadcrumbSeparator spacing={spacing} children={separator} />
      )}
    </chakra.li>
  )
})

if (__DEV__) {
  BreadcrumbItem.displayName = "BreadcrumbItem"
}

export interface BreadcrumbOptions {
  children: ReactNode
  /**
   * The visual separator between each breadcrumb item
   */
  separator?: string | ReactElement
  /**
   * The left and right margin applied to the separator
   */
  spacing?: SystemProps["mx"]
}

export type BreadcrumbProps = Omit<PropsOf<typeof chakra.nav>, "children"> &
  BreadcrumbOptions &
  ThemingProps & { children: ReactNode }

/**
 * Breadcrumb is used to render a breadcrumb navigation landmark.
 * It renders a `nav` element with `aria-label` set to `Breadcrumb`
 *
 * @see Docs https://chakra-ui.com/components/breadcrumbs
 */
export const Breadcrumb = forwardRef(function Breadcrumb(
  props: BreadcrumbProps,
  ref: Ref<HTMLElement>,
) {
  const styles = useMultiStyleConfig("Breadcrumb", props)
  const realProps = omitThemingProps(props)

  const {
    children,
    spacing = "0.5rem",
    separator = "/",
    className,
    ...rest
  } = realProps

  const validChildren = getValidChildren(children)
  const count = validChildren.length

  const clones = validChildren.map((child, index) =>
    cloneElement(child, {
      separator,
      spacing,
      isLastChild: count === index + 1,
    }),
  )

  const _className = cx("chakra-breadcrumb", className)

  return (
    <chakra.nav
      ref={ref}
      aria-label="breadcrumb"
      className={_className}
      {...rest}
    >
      <StylesProvider value={styles}>
        <chakra.ol className="chakra-breadcrumb__list">{clones}</chakra.ol>
      </StylesProvider>
    </chakra.nav>
  )
})

if (__DEV__) {
  Breadcrumb.displayName = "Breadcrumb"
}
