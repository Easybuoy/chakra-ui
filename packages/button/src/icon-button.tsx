import { __DEV__ } from "@chakra-ui/utils"
import React, {
  ReactElement,
  isValidElement,
  cloneElement,
  Ref,
  ReactNode,
} from "react"
import { Button, ButtonProps } from "./button"
import { forwardRef } from "@chakra-ui/system"

type Omitted =
  | "leftIcon"
  | "isFullWidth"
  | "rightIcon"
  | "loadingText"
  | "children"

type BaseButtonProps = Omit<ButtonProps, Omitted> & { children?: ReactNode }

export type IconButtonProps = BaseButtonProps & {
  icon?: ReactElement
  isRound?: boolean
  "aria-label": string
}

export const IconButton = forwardRef(function IconButton(
  props: IconButtonProps,
  ref: Ref<HTMLButtonElement>,
) {
  const { icon, children, isRound, "aria-label": ariaLabel, ...rest } = props

  /**
   * Passing the icon as prop or children should work
   */
  const element = icon || children
  const _children = isValidElement(element)
    ? cloneElement(element as any, {
        "aria-hidden": true,
        focusable: false,
      })
    : null

  return (
    <Button
      padding="0"
      borderRadius={isRound ? "full" : "md"}
      ref={ref}
      aria-label={ariaLabel}
      {...rest}
    >
      {_children}
    </Button>
  )
})

if (__DEV__) {
  IconButton.displayName = "IconButton"
}
