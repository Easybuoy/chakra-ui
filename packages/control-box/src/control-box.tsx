import {
  chakra,
  SystemStyleObject,
  PropsOf,
  forwardRef,
} from "@chakra-ui/system"
import { __DEV__ } from "@chakra-ui/utils"
import React, { ReactNode, Ref } from "react"

export type ControlBoxOptions = {
  type?: "checkbox" | "radio"
  _hover?: SystemStyleObject
  _invalid?: SystemStyleObject
  _disabled?: SystemStyleObject
  _focus?: SystemStyleObject
  _checked?: SystemStyleObject
  _child?: SystemStyleObject
  _checkedAndChild?: SystemStyleObject
  _checkedAndDisabled?: SystemStyleObject
  _checkedAndFocus?: SystemStyleObject
  _checkedAndHover?: SystemStyleObject
}

export type ControlBoxProps = Omit<PropsOf<typeof chakra.div>, "children"> &
  ControlBoxOptions & { children: ReactNode }

export const ControlBox = forwardRef(function ControlBox(
  props: ControlBoxProps,
  ref: Ref<HTMLDivElement>,
) {
  const {
    type = "checkbox",
    _hover,
    _invalid,
    _disabled,
    _focus,
    _checked,
    _child = { opacity: 0 },
    _checkedAndChild = { opacity: 1 },
    _checkedAndDisabled,
    _checkedAndFocus,
    _checkedAndHover,
    children,
    ...rest
  } = props

  const checkedAndDisabled = `input[type=${type}]:checked:disabled + &`
  const checkedAndHover = `input[type=${type}]:checked:hover:not(:disabled) + &`
  const checkedAndFocus = `input[type=${type}]:checked:focus + &`
  const disabled = `input[type=${type}]:disabled + &`
  const focus = `input[type=${type}]:focus + &`
  const hover = `input[type=${type}]:hover:not(:disabled):not(:checked) + &`
  const checked = `input[type=${type}]:checked + &, input[type=${type}][aria-checked=mixed] + &`
  const invalid = `input[type=${type}][aria-invalid=true] + &`
  const child = `& > *`

  return (
    <chakra.div
      ref={ref}
      {...rest}
      aria-hidden
      __css={{
        display: "inline-flex",
        alignItems: "center",
        justifyContent: "center",
        transition: "all 120ms",
        flexShrink: "0",
        [focus]: _focus,
        [hover]: _hover,
        [disabled]: _disabled,
        [invalid]: _invalid,
        [checkedAndDisabled]: _checkedAndDisabled,
        [checkedAndFocus]: _checkedAndFocus,
        [checkedAndHover]: _checkedAndHover,
        [child]: _child,
        [checked]: {
          ..._checked,
          [child]: _checkedAndChild,
        },
      }}
    >
      {children}
    </chakra.div>
  )
})

if (__DEV__) {
  ControlBox.displayName = "ControlBox"
}
