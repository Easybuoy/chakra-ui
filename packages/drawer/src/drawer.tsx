import React, {
  createContext,
  useContext,
  ReactNode,
  CSSProperties,
  Ref,
} from "react"
import { Slide, SlideProps, Fade } from "@chakra-ui/transition"
import {
  Modal,
  ModalProps,
  ModalContentProps,
  ModalContent,
  ModalOverlayProps,
  ModalOverlay,
  ModalBody,
  ModalHeader,
  ModalFooter,
  ModalCloseButton,
} from "@chakra-ui/modal"
import { forwardRef } from "@chakra-ui/system"
import { __DEV__ } from "@chakra-ui/utils"

interface TransitionStyles {
  content: CSSProperties
  overlay: CSSProperties
}

const TransitionContext = createContext<TransitionStyles>({
  content: {},
  overlay: {},
})

TransitionContext.displayName = "TransitionContext"
const useTransitionContext = () => {
  const context = useContext(TransitionContext)

  if (!context) {
    throw new Error(
      "useTransitionContext() was called outside of a TransitionContext.Provider in a Drawer component\nYou probably forgot to use `Drawer`.",
    )
  }

  return context
}

interface DrawerTransitionProps {
  in: boolean
  children: ReactNode
  placement: SlideProps["placement"]
}

function DrawerTransition(props: DrawerTransitionProps) {
  const { in: inProp, children, placement } = props
  return (
    <Slide
      timeout={250}
      in={inProp}
      placement={placement}
      transition="transform 250ms cubic-bezier(0.16, 1, 0.3, 1)"
    >
      {(content) => (
        <Fade
          timeout={250}
          in={inProp}
          transition="opacity 250ms cubic-bezier(0.16, 1, 0.3, 1)"
        >
          {(overlay) => (
            <TransitionContext.Provider value={{ content, overlay: overlay }}>
              {children}
            </TransitionContext.Provider>
          )}
        </Fade>
      )}
    </Slide>
  )
}

if (__DEV__) {
  DrawerTransition.displayName = "DrawerTransition"
}

export interface DrawerProps extends ModalProps {
  placement?: SlideProps["placement"]
  isFullHeight?: boolean
}

export function Drawer(props: DrawerProps) {
  const { isOpen, onClose, placement = "right", children, ...rest } = props
  return (
    <DrawerTransition in={isOpen} placement={placement}>
      <Modal isOpen onClose={onClose} {...rest}>
        {children}
      </Modal>
    </DrawerTransition>
  )
}

export const DrawerContent = forwardRef(function DrawerContent(
  props: ModalContentProps,
  ref: Ref<HTMLDivElement>,
) {
  const { content: styles } = useTransitionContext()
  return (
    <ModalContent
      ref={ref}
      position="fixed"
      style={styles}
      marginTop="0"
      marginBottom="0"
      borderRadius="0"
      {...props}
    />
  )
})

export const DrawerOverlay = forwardRef(function DrawerOverlay(
  props: ModalOverlayProps,
  ref: Ref<HTMLDivElement>,
) {
  const { overlay: styles } = useTransitionContext()
  return <ModalOverlay style={styles} ref={ref} {...props} />
})

export const DrawerBody = ModalBody
export const DrawerHeader = ModalHeader
export const DrawerFooter = ModalFooter
export const DrawerCloseButton = ModalCloseButton
