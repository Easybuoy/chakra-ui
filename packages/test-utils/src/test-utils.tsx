import theme from "@chakra-ui/theme"
import { GlobalStyle, ThemeProvider } from "@chakra-ui/system"
import CSSReset from "@chakra-ui/css-reset"
import "@testing-library/jest-dom/extend-expect"
import { render, RenderOptions, fireEvent } from "@testing-library/react"
import React, { ReactElement } from "react"
import { toHaveNoViolations, axe } from "jest-axe"
import serializer from "jest-emotion"

expect.addSnapshotSerializer(serializer)

expect.extend(toHaveNoViolations)

const AllProviders = ({ children }: { children?: React.ReactNode }) => (
  <ThemeProvider theme={theme}>
    <CSSReset />
    <GlobalStyle />
    {children}
  </ThemeProvider>
)

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllProviders, ...options })

/**
 * Wrapper for jest-axe
 *
 * @see https://github.com/nickcolley/jest-axe#testing-react-with-react-testing-library
 */
async function testA11Y(component: ReactElement, options?: RenderOptions) {
  const { container } = render(component, options)

  const results = await axe(container)

  expect(results).toHaveNoViolations()
}

export * from "@testing-library/react"

export {
  act as invoke,
  renderHook,
  RenderHookOptions,
  RenderHookResult,
} from "@testing-library/react-hooks"

export { default as userEvent } from "@testing-library/user-event"

export { customRender as render, testA11Y }

export * from "jest-axe"

export const escape = (ui: HTMLElement) =>
  fireEvent.keyDown(ui, { key: "Escape", keyCode: 27 })
