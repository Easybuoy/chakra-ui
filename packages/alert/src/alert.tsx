import { CheckIcon, InfoIcon, WarningIcon } from "./icons"
import {
  chakra,
  PropsOf,
  StylesProvider,
  useStyles,
  ThemingProps,
  useMultiStyleConfig,
} from "@chakra-ui/system"
import { createContext, cx } from "@chakra-ui/utils"
import React, { forwardRef, Ref, ReactNode } from "react"
import { DivProps } from "./localTypes"

export const STATUSES = {
  info: { icon: InfoIcon, colorScheme: "blue" },
  warning: { icon: WarningIcon, colorScheme: "orange" },
  success: { icon: CheckIcon, colorScheme: "green" },
  error: { icon: WarningIcon, colorScheme: "red" },
}

type AlertContext = { status: keyof typeof STATUSES }

const [AlertProvider, useAlertContext] = createContext<AlertContext>({
  name: "AlertContext",
  errorMessage:
    "useAlertContext: `context` is undefined. Seems you forgot to wrap alert components in `<Alert />`",
})

interface AlertOptions {
  /**
   * The status of the alert
   */
  status?: keyof typeof STATUSES
}

export type AlertProps = Omit<DivProps, "children"> &
  AlertOptions &
  ThemingProps & { children: ReactNode }

/**
 * Alert
 *
 * React component used to communicate the state or status of a
 * page, feature or action
 */
export const Alert = forwardRef(function Alert(
  props: AlertProps,
  ref: Ref<HTMLDivElement>,
) {
  const { status = "info", ...rest } = props
  const { colorScheme } = STATUSES[status]

  const styles = useMultiStyleConfig("Alert", { ...props, colorScheme })
  const alertStyles = {
    width: "100%",
    display: "flex",
    alignItems: "center",
    position: "relative",
    overflow: "hidden",
    ...styles.container,
  }

  return (
    <AlertProvider value={{ status }}>
      <StylesProvider value={styles}>
        <chakra.div
          role="alert"
          ref={ref}
          {...rest}
          className={cx("chakra-alert", props.className)}
          __css={alertStyles}
        />
      </StylesProvider>
    </AlertProvider>
  )
})

export type AlertTitleProps = Omit<DivProps, "children"> & {
  children: ReactNode
}

export const AlertTitle = forwardRef(function AlertTitle(
  props: AlertTitleProps,
  ref: Ref<HTMLDivElement>,
) {
  const styles = useStyles()
  return (
    <chakra.div
      ref={ref}
      {...props}
      className={cx("chakra-alert__title", props.className)}
      __css={styles.title}
    />
  )
})

export type AlertDescriptionProps = Omit<DivProps, "children"> & {
  children: ReactNode
}

export const AlertDescription = forwardRef(function AlertDescription(
  props: AlertDescriptionProps,
  ref: Ref<HTMLDivElement>,
) {
  const styles = useStyles()
  const descriptionStyles = {
    display: "inline-block",
    ...styles.description,
  }

  return (
    <chakra.div
      ref={ref}
      {...props}
      className={cx("chakra-alert__desc", props.className)}
      __css={descriptionStyles}
    />
  )
})

export type AlertIconProps = Omit<PropsOf<typeof chakra.span>, "children">

export const AlertIcon = (props: AlertIconProps) => {
  const { status } = useAlertContext()
  const { icon: BaseIcon } = STATUSES[status]
  const styles = useStyles()

  return (
    <chakra.span
      display="inherit"
      {...props}
      className={cx("chakra-alert__icon", props.className)}
      __css={styles.icon}
    >
      <BaseIcon w="100%" h="100%" />
    </chakra.span>
  )
}
