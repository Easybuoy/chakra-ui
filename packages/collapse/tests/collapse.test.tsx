import React from "react"
import { render, testA11Y } from "@chakra-ui/test-utils"
import { Collapse } from "../src"

describe("<Collapse />", () => {
  test("renders correctly", () => {
    const { asFragment } = render(<Collapse>sfsjdfkbfkjdfdjksf</Collapse>)
    expect(asFragment()).toMatchSnapshot()
  })

  test("passes a11y test", async () => {
    await testA11Y(<Collapse>sfsjdfkbfkjdfdjksf</Collapse>)
  })
})
