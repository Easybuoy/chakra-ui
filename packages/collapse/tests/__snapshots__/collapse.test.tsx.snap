// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Collapse /> renders correctly 1`] = `
<DocumentFragment>
  <div
    aria-hidden="true"
    class="rah-static rah-static--height-zero"
    style="height: 0px; overflow: hidden;"
    sx="[object Object]"
  >
    <div
      style="transition: opacity 250ms ease 0ms; opacity: 0; display: none;"
    >
      <div
        class="css-0"
      >
        sfsjdfkbfkjdfdjksf
      </div>
    </div>
  </div>
</DocumentFragment>
`;
